src/Interpreters/Cache/Metadata.cpp

```diff
@@ -938,8 +938,18 @@ KeyMetadata::iterator LockedKey::removeFileSegmentImpl(
     try
     {
         const auto path = key_metadata->getFileSegmentPath(*file_segment);
-        bool exists = fs::exists(path);
-        if (exists)
+        if (file_segment->segment_kind == FileSegmentKind::Temporary)
+        {
+            /// FIXME: For temporary file segment the requirement is not as strong because
+            /// the implementation of "temporary data in cache" creates files in advance.
+            if (fs::exists(path))
+                fs::remove(path);
+        }
+        else if (file_segment->downloaded_size == 0)
+        {
+            chassert(!fs::exists(path));
+        }
+        else if (fs::exists(path))
```

Comment:
The point of this PR is not to do `fs::exists` check unconditionally, but only if `downloaded_size == 0`.

Reply:
> The point of this PR is not to do `fs::exists` check unconditionally, but only if `downloaded_size == 0`.

you mean `!=`?

Reply:
Oops, yes, typo